articles/qos_deadline_liveliness_lifespan.md

```diff
@@ -25,102 +25,155 @@ Glossary:
 - DDS - Data Distribution Service
 - RTPS - Real-Time Publish Subscribe
 - QoS - Quality of Service
+- Client - Refers to an application that connects to a ROS Service to send requests and receive responses.
+- Owner - Refers to the application that is running a ROS Service that receives requests and sends responses.
 
 ## Existing ROS QoS Settings
 
-While DDS provides many settings to enable fine-grained control over the Quality of Service (QOS) for entities, ROS only provides native support for a handful of them. ROS users are able to specify the History, Depth, Reliability, and Durability via a QoS configuration struct when they create Publishers, Subscribers, etc.
+While DDS provides many settings to enable fine-grained control over the Quality of Service (QoS) for entities, ROS only provides native support for a handful of them. ROS users are able to specify the History, Depth, Reliability, and Durability via a QoS configuration struct when they create Publishers, Subscribers, etc.
 
 This leaves a lot of QoS settings that can only be set if DDS vendor can load additional default settings via a configuration file.
 If a user wants to hook their code into these additional QoS settings then they would need to get a reference to the rmw implementation and program against the vendor specific API.
 Without the abstraction layer provided by ROS their code becomes less portable.
 
-## Proposed Changes
+## New QoS Settings
 
 As users start to build more robust applications, they are going to need more control over when data is being delivered and information about when the system is failing.
 To address these needs it was proposed that we start by adding support for the following new DDS QoS settings.
 
-### Proposed New QoS Settings
+### Deadline
+
+The deadline policy establishes a contract for the amount of time allowed between messages.
+For Topic Subscribers it establishes the maximum amount of time allowed to pass between receiving messages.
+For Topic Publishers it establishes the maximum amount of time allowed to pass between sending messages.
+For Service Owners it establishes the maximum amount of time allowed to pass between receiving a request and when a response for that request is sent.
+For Service Clients it establishes the maximum amount of time allowed to pass between sending a request and when a response for that request is received.
+
```

Comment:
We can leave a sentence or two explaining that actions are built on top of services and topics, and therefore the above policies apply to action servers and action clients accordingly, ie, action servers are service owners and topic publishers; action clients are service clients and topic subscribers.

Reply:
If that works for actions, then yes, I think we should do that. But... does it work for actions? (Sorry, haven't had the cycles to sit down and think about it properly.) One of my main concerns is it makes it explicit that actions are a combination of services and topics, rather than something in their own right.

Reply:
TBH, I didn't invest too much thought either, but I don't foresee any issues with extending the definitions to the components that make up actions. Although, I think it would be worth noting what their default settings are. Two components that stand out are the result service and status topic. It's easy to see how the deadline QoS setting would be useful for the result service (e.g. if the application developer has some expected upper bound on completion time). But, since time between a result request and the response is application dependent, we probably want to default to a "no deadline" setting. The status topic updates only when a goal state changes, so it seems the deadline setting is not very useful and should also be set to "no deadline", but I don't see any harm in letting the user configure it as they please.

If we want to think about actions as a black box (making no implementation assumptions about goal requests, results, feedback, etc), then I think it makes sense to discuss what *deadline* and *liveness* mean for the five major parts of an action: goal, result, cancel request, feedback, and status. But given that the action design doc describes actions in terms of services and topics, I'm not sure if there is much value in trying to make this abstraction here.

Reply:
I can see the usefulness of trying to define what these QoS policies might mean for actions, like maybe a deadline is the amount of time an action should be completed in or how often it needs to provide updates. My hesitance to do so is because of the fact that they are exposing in their design and in their API the underlying topics/services. If an action level QoS settings requires a change to the underlying Topic or Service QoS settings then you'd have to override what the user has provided, which isn't the best api experience. 

I don't think anything defined here limits designing additional QoS policies for actions in the future. However, not explicitly including them was useful in keeping the design and subsequent review small. I think a future review could focus just on defining additional QoS settings for actions and do it more justice than I can at the moment trying to fit them into this design.

Reply:
I think in terms of action behaviour, deadline makes the most sense to me for the status update. The result is very dependent on the performance of the action, but there will be cases where I want to know within a deadline that the action is still proceeding as planned, or has had an error, or whatever.

Deadline is also important for knowing in a timely manner that an action has been accepted for execution, or that it has been rejected.

Liveliness makes sense for the action as a whole: Is the action server still alive and responsive? Is the client still around to care if the action finishes or not?

There was some talk about when a client can query the result and how long the action server needs to hold on to results for in case the client asks about it. Lifespan is relevant to this.

Probably the best approach is to list each piece of information that can be communicated between the action client and action server and look at how each of the three QoS policies can be used for them, if at all.

Reply:
@gbiggs I might very well be wrong here but my understanding is that all three OoS discussed here apply to the Data Readers and Data Writers. So the implement something like this:

> Liveliness makes sense for the action as a whole: Is the action server still alive and responsive? Is the client still around to care if the action finishes or not?

we would need to first determine a) the current computational state of the ROS 2 node and b) the current ROS 2 node level health state (OK, WARN, ERROR, STALE) and then publish this separately. What I think would probably make most sense is to create a separate topic for such information and set `liveliness` for it. At which point this would be ROS 2 node and its derivatives dependent and not bound to this discussion here.

Reply:
@dejanpan The three QoS policies discussed here derive from the DDS spec for DataReaders/Writers, but they are defined here in terms of the behavior of the ROS primitives. So where it makes sense we can just leverage the DDS settings for implementation and where it doesn't they can be implemented separate. I don't think there are any technical blockers for adding these concepts to Actions, I just hadn't been focused on defining them for Actions.